src/pipdeptree/_discovery.py

```diff
@@ -1,35 +1,39 @@
 from __future__ import annotations
 
+import site
+import sys
+from importlib.metadata import distributions
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from pip._vendor.pkg_resources import DistInfoDistribution
+    from importlib.metadata import Distribution
+
+
+def is_virtual_environment() -> bool:
+    return sys.prefix != sys.base_prefix
+
+
+def get_site_packages_directory() -> list[str] | None:
+    if is_virtual_environment():
+        return site.getsitepackages()
+    return None
 
 
 def get_installed_distributions(
     local_only: bool = False,  # noqa: FBT001, FBT002
     user_only: bool = False,  # noqa: FBT001, FBT002
-) -> list[DistInfoDistribution]:
-    try:
-        from pip._internal.metadata import pkg_resources  # noqa: PLC0415, PLC2701
-    except ImportError:
-        # For backward compatibility with python ver. 2.7 and pip
-        # version 20.3.4 (the latest pip version that works with python
-        # version 2.7)
-        from pip._internal.utils import misc  # noqa: PLC0415, PLC2701 # pragma: no cover
-
-        return misc.get_installed_distributions(  # type: ignore[no-any-return,attr-defined]
-            local_only=local_only,
-            user_only=user_only,
-        )
-
-    else:
-        dists = pkg_resources.Environment.from_paths(None).iter_installed_distributions(
-            local_only=local_only,
-            skip=(),
-            user_only=user_only,
-        )
-        return [d._dist for d in dists]  # type: ignore[attr-defined] # noqa: SLF001
+) -> list[Distribution]:
+    dists = distributions()
+
+    if local_only:
+        site_packages = get_site_packages_directory() or []
+        dists = [d for d in dists if str(d.locate_file("")) in site_packages]
+
+    if user_only:
+        user_site = site.getusersitepackages() or []
+        dists = [d for d in dists if str(d.locate_file("")) == user_site]
+
+    return dists
```

Comment:
Let's do this instead:

```suggestion
    if user_only:
        return list(distributions(path=[site.getusersitepackages()]))

    # NOTE: See https://docs.python.org/3/library/venv.html#how-venvs-work for more details.
    in_venv = sys.prefix != sys.base_prefix

    if local_only and in_venv:
        # TODO: Are venvs given only one site package?
        venv_site_packages = [sp for sp in site.getsitepackages() or [] if sp.startswith(sys.prefix)]
        return list(distributions(path=venv_site_packages))

    return list(distributions())
```